From 52909131ab4dd26eca73dcccb33afddadbf67765 Mon Sep 17 00:00:00 2001
From: Hannes Reinecke <hare@suse.de>
Date: Thu, 13 Oct 2016 15:10:44 +0200
Subject: scsi: libfc: use error code for fc_rport_error()
Git-commit: 9f9504a7cdee39e167f0421346ff17568a5f29a0
Patch-mainline: v4.10-rc1
Git-commit: 52909131ab4dd26eca73dcccb33afddadbf67765
References: bsc#1023764

We only ever use the 'fp' argument for fc_rport_error() to
encapsulate the error code, so we can as well do away with that
and pass the error directly.

Signed-off-by: Hannes Reinecke <hare@suse.com>
Acked-by: Johannes Thumshirn <jth@kernel.org>
Signed-off-by: Martin K. Petersen <martin.petersen@oracle.com>
Signed-off-by: Johannes Thumshirn <jthumshirn@suse.de>

---
 drivers/scsi/libfc/fc_rport.c | 17 +++++++++++------
 1 file changed, 11 insertions(+), 6 deletions(-)

diff --git a/drivers/scsi/libfc/fc_rport.c b/drivers/scsi/libfc/fc_rport.c
index 7907c27..3a56160 100644
--- a/drivers/scsi/libfc/fc_rport.c
+++ b/drivers/scsi/libfc/fc_rport.c
@@ -741,10 +741,11 @@ static void fc_rport_flogi_resp(struct fc_seq *sp, struct fc_frame *fp,
 	struct fc_lport *lport = rdata->local_port;
 	struct fc_els_flogi *flogi;
 	unsigned int r_a_tov;
+	u8 opcode;
 	int err = 0;
 
 	FC_RPORT_DBG(rdata, "Received a FLOGI %s\n",
-		     IS_ERR(fp)? "error" : fc_els_resp_type(fp));
+		     IS_ERR(fp) ? "error" : fc_els_resp_type(fp));
 
 	if (fp == ERR_PTR(-FC_EX_CLOSED))
 		goto put;
@@ -763,15 +764,19 @@ static void fc_rport_flogi_resp(struct fc_seq *sp, struct fc_frame *fp,
 		fc_rport_error(rdata, PTR_ERR(fp));
 		goto err;
 	}
-
-	if (fc_frame_payload_op(fp) != ELS_LS_ACC) {
+	opcode = fc_frame_payload_op(fp);
+	if (opcode == ELS_LS_RJT) {
 		struct fc_els_ls_rjt *rjt;
 
-		rjt = fc_frame_payload_get(fp, sizeof (*rjt));
+		rjt = fc_frame_payload_get(fp, sizeof(*rjt));
 		FC_RPORT_DBG(rdata, "FLOGI ELS rejected, reason %x expl %x\n",
 			     rjt->er_reason, rjt->er_explan);
 		err = -FC_EX_ELS_RJT;
 		goto bad;
+	} else if (opcode != ELS_LS_ACC) {
+		FC_RPORT_DBG(rdata, "FLOGI ELS invalid opcode %x\n", opcode);
+		err = -FC_EX_ELS_RJT;
+		goto bad;
 	}
 	if (fc_rport_login_complete(rdata, fp)) {
 		FC_RPORT_DBG(rdata, "FLOGI failed, no login\n");
@@ -781,7 +786,6 @@ static void fc_rport_flogi_resp(struct fc_seq *sp, struct fc_frame *fp,
 
 	flogi = fc_frame_payload_get(fp, sizeof(*flogi));
 	if (!flogi) {
-		FC_RPORT_DBG(rdata, "Bad FLOGI response\n");
 		err = -FC_EX_ALLOC_ERR;
 		goto bad;
 	}
@@ -801,6 +805,7 @@ put:
 	kref_put(&rdata->kref, lport->tt.rport_destroy);
 	return;
 bad:
+	FC_RPORT_DBG(rdata, "Bad FLOGI response\n");
 	fc_rport_error_retry(rdata, err);
 	goto out;
 }
@@ -1030,7 +1035,7 @@ static void fc_rport_plogi_resp(struct fc_seq *sp, struct fc_frame *fp,
 	} else {
 		struct fc_els_ls_rjt *rjt;
 
-		rjt = fc_frame_payload_get(fp, sizeof (*rjt));
+		rjt = fc_frame_payload_get(fp, sizeof(*rjt));
 		FC_RPORT_DBG(rdata, "PLOGI ELS rejected, reason %x expl %x\n",
 			     rjt->er_reason, rjt->er_explan);
 		fc_rport_error_retry(rdata, -FC_EX_ELS_RJT);
-- 
1.8.5.6

